File: hashMap.py
'''
time complexity: O(n)
space complexity: O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
'''
class Node:
    def __init__(self,key,value):
        self.key = key
        self.value = value
        self.next = None
class MyHashMap:

    def __init__(self):
        self.bucket = 1000
        self.storage = [None] * self.bucket 
        
    def bucketHash(self,key: int) -> int:
        return key % self.bucket
    
    
    def find(self,head,key):

        prev = None
        curr = head
        while(curr!=None and curr.key!=key):
            prev = curr
            curr = curr.next
        return prev
        
    def put(self, key: int, value: int) -> None:
        hashValue = self.bucketHash(key)
        if(self.storage[hashValue]==None):
            self.storage[hashValue] = Node(-1,-1)
        prev = self.find(self.storage[hashValue],key)
        if(prev.next==None):
            prev.next = Node(key,value)
        else:
            prev.next.value = value

    def get(self, key: int) -> int:
        hashValue = self.bucketHash(key)
        if(self.storage[hashValue]==None): return -1
        prev = self.find(self.storage[hashValue],key)
        if(prev.next==None): return -1
        return prev.next.value
        

    def remove(self, key: int) -> None:
        hashValue = self.bucketHash(key)
        if(self.storage[hashValue]==None): return
        prev = self.find(self.storage[hashValue],key)
        if(prev.next==None): return 
        prev.next = prev.next.next


# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)

File: Hashmap.cpp
// Time Complexity :T(n) = All operations are O(1)
// Space Complexity :S(n) = O(n)
// Did this code successfully run on Leetcode : yes


// Your code here along with comments explaining your approach

// used double hashing to implement hashmap


class MyHashMap {
public:
    /** Initialize your data structure here. */
     vector<vector<int>> map;
     int buckets;
     int bucketitems;
    
    MyHashMap() {
        this->map.resize(1000);
        this->buckets=1000;
        this->bucketitems=1000;
    }
     int getBuckets(int key)
    {
        return key%buckets;
    }
    int getBucketItems(int key)
    {
        return key/bucketitems;
    }
    
    /** value will always be non-negative. */
    void put(int key, int value) {
        
        int Bucket=getBuckets(key);
        int BucketItem=getBucketItems(key);
        if(map[Bucket].size() < BucketItem+1)
        {
            this->map[Bucket].resize(BucketItem+1,-1);   
        }
        this->map[Bucket][BucketItem]=value;
             
    }
    
    /** Returns the value to which the specified key is mapped, or -1 if this map contains no mapping for the key */
    int get(int key) 
    {
        
        int Bucket=getBuckets(key);
        int BucketItem=getBucketItems(key);
        
        if (map[Bucket].size()<BucketItem+1) 
        {
            return -1;
        }
        else
        {
                 if (map[Bucket][BucketItem]==-1)
                {
                    return -1;
                }
                else
                {
                    return map[Bucket][BucketItem];
                }
        }
        
        
    }
    
    /** Removes the mapping of the specified value key if this map contains a mapping for the key */
    void remove(int key) {
        int Bucket=getBuckets(key);
        int BucketItem=getBucketItems(key);
        
        if (map[Bucket].size()<BucketItem+1) 
        {
            return;
        }
        map[Bucket][BucketItem]=-1;
    }
};

/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap* obj = new MyHashMap();
 * obj->put(key,value);
 * int param_2 = obj->get(key);
 * obj->remove(key);
 */
